Fix and modernise touch events example

hamishwillee opened the pull request:
The [Touch events](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events#results) example did not work. This fixes it as suggested by @wbamberg in https://github.com/mdn/content/issues/13590#issuecomment-1064765288

- [x] fix the argument to EmbedLiveSample and fix the size of the iframe so it's big enough for the example
- [x] remove the LiveSampleLink macro call and the JSFiddle link
- [x] update the example to have an "Initialize" button, like the JSFiddle version
- [x] update the code to be modern JS style
- [x] add a note about how to enable touch event simulation for desktop browsers

Fixes #13590

@wbamberg The example works. I "think" I got the const/let right :-)

hamishwillee committed:
Fix and modernise touch events example

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -83,14 +86,14 @@ When a {{domxref("Element/touchstart_event", "touchstart")}} event occurs, indic
+  const touches = evt.changedTouches;
 
-  for (var i = 0; i < touches.length; i++) {
+  for (let i = 0; i < touches.length; i++) {
```
Also, there's a lot of code in here to handle multitouch, but I can't get this to work. Is that a hardware thing?

hamishwillee replied:
@wbamberg I think it is worth keeping the console logging code - it is useful/interesting to watch in desktop. Whereas I don't like the in-page log because it doesn't scroll/expand nicely. I think I'll leave this bit alone - if you feel strongly, then perhaps as a post process.

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -111,13 +114,13 @@ Each time one or more fingers move, a {{domxref("Element/touchmove_event", "touc
-  for (var i = 0; i < touches.length; i++) {
-    var color = colorForTouch(touches[i]);
-    var idx = ongoingTouchIndexById(touches[i].identifier);
+  for (let i = 0; i < touches.length; i++) {
```
```suggestion
  for (const touch of touches) {
```

(plus all the corresponding changes inside the loop)

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -111,13 +114,13 @@ Each time one or more fingers move, a {{domxref("Element/touchmove_event", "touc
-    var idx = ongoingTouchIndexById(touches[i].identifier);
+  for (let i = 0; i < touches.length; i++) {
+    const color = colorForTouch(touches[i]);
+    let idx = ongoingTouchIndexById(touches[i].identifier);
```
```suggestion
    const idx = ongoingTouchIndexById(touches[i].identifier);
```

wbamberg replied:
You can also replace the `ongoingTouchIndexById` function with the one-liner:

```js
ongoingTouches.findIndex(ongoingTouch => ongoingTouch.identifier === touch.identifier);
```

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -153,13 +156,13 @@ When the user lifts a finger off the surface, a {{domxref("Element/touchend_even
-  for (var i = 0; i < touches.length; i++) {
-    var color = colorForTouch(touches[i]);
-    var idx = ongoingTouchIndexById(touches[i].identifier);
+  for (let i = 0; i < touches.length; i++) {
```
```suggestion
  for (const touch of touches) {
```

hamishwillee replied:
As above, I'd rather keep the console logging.

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -153,13 +156,13 @@ When the user lifts a finger off the surface, a {{domxref("Element/touchend_even
-    var idx = ongoingTouchIndexById(touches[i].identifier);
+  for (let i = 0; i < touches.length; i++) {
+    const color = colorForTouch(touches[i]);
+    let idx = ongoingTouchIndexById(touches[i].identifier);
```
...and the same comment here as before

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -186,10 +189,10 @@ If the user's finger wanders into browser UI, or the touch otherwise needs to be
 
-  for (var i = 0; i < touches.length; i++) {
-    var idx = ongoingTouchIndexById(touches[i].identifier);
+  for (let i = 0; i < touches.length; i++) {
```
```suggestion
  for (const touch of touches) {
```

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -186,10 +189,10 @@ If the user's finger wanders into browser UI, or the touch otherwise needs to be
-  for (var i = 0; i < touches.length; i++) {
-    var idx = ongoingTouchIndexById(touches[i].identifier);
+  for (let i = 0; i < touches.length; i++) {
+    let idx = ongoingTouchIndexById(touches[i].identifier);
```
Also same comment as beforew.

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -207,13 +210,13 @@ To make each touch's drawing look different, the `colorForTouch()` function is u
   g = g.toString(16); // make it a hex digit
   b = b.toString(16); // make it a hex digit
-  var color = "#" + r + g + b;
+  let color = "#" + r + g + b;
```
```suggestion
  const color = "#" + r + g + b;
```

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -252,16 +255,21 @@ function ongoingTouchIndexById(idToFind) {
 
-{{EmbedLiveSample('Demo')}}
+> **Note:** More generally, the example will work on platforms that provide touch events. 
+> You can test this on desktop versions of Firefox by enabling "touch simulation" in [Responsive Design Mode](/en-US/docs/Tools/Responsive_Design_Mode#toggling_responsive_design_mode), and then pressing the "Initialize" button to reload the content.
```
I think we should also point to https://developer.chrome.com/docs/devtools/device-mode/, which seems to be Chrome's version, and which I've confirmed works.

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -252,16 +255,21 @@ function ongoingTouchIndexById(idToFind) {
+> You can test this on desktop versions of Firefox by enabling "touch simulation" in [Responsive Design Mode](/en-US/docs/Tools/Responsive_Design_Mode#toggling_responsive_design_mode), and then pressing the "Initialize" button to reload the content.
 
-You can also {{LiveSampleLink('Demo', 'view the results of the demo code in a separate page')}} (or try it out on [jsFiddle here](https://jsfiddle.net/Darbicus/z3Xdx/10/)).
+You can also [try it out on jsFiddle here](https://jsfiddle.net/Darbicus/z3Xdx/10/).
```
I would remove this link. We don't control the contents of it and I don't think it adds anything.

wbamberg commented on files/en-us/web/api/touch_events/index.md:
```diff
@@ -277,9 +285,9 @@ function onTouch(evt) {
-  var newEvt = document.createEvent("MouseEvents");
-  var type = null;
-  var touch = null;
+  let newEvt = document.createEvent("MouseEvents");
```
```suggestion
  const newEvt = document.createEvent("MouseEvents");
```

wbamberg reviewed (dismissed):
Thanks for the update Hamish!

This was a bit of a "surface" review, I didn't really dig into the logic of it. But it seems to work fine as far as I could test it, and it is quite a neat example. And it is a lot better than it was!

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented on files/en-us/web/api/touch_events/index.md:
````diff
@@ -57,11 +59,12 @@ When the page loads, the `startup()` function shown below will be called.
 ```js
 function startup() {
-  var el = document.getElementById("canvas");
+  const el = document.getElementById("canvas");
````
Why? Both seem to work and MDN is completely inconsistent.

wbamberg replied:
Fair enough. I always try to use single quotes, and it's what our guidelines say: https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/JavaScript#use_expanded_syntax. But there's not a lot of value in trying to make this happen manually, we need Prettier for this kind of thing.

hamishwillee replied:
Thanks! I've done it anyway, just wanted to confirm reason. If that's what guidelines and prettier does, then great.

hamishwillee committed:
Make strings and interpolation into modern format

hamishwillee committed:
Remove explicit false on event handlers

hamishwillee committed:
Add link for chrome dev tools

hamishwillee committed:
Use listener for the button

hamishwillee commented:
@wbamberg OK, I have fixed the bits that I want to :-)

The console logging and in-page logging are inconsistent (as you point out). I would be happy to change all the console logging to go into the in-page log if you like or move everything into the console (and recommend users look at console logging on desktop, which is where most developers are).

There must be a way to make the log fixed length and scroll, so I'd do that.

Happy to take direction on which, but I don't want to strip the logging out.

I am slightly tempted to remove the intialize button too. We added it to make simulation case easier - but you don't "need" it for the example, and it isn't hard to say "just reload the page".

wbamberg commented:
> @wbamberg OK, I have fixed the bits that I want to :-)
> 
> The console logging and in-page logging are inconsistent (as you point out). I would be happy to change all the console logging to go into the in-page log if you like or move everything into the console (and recommend users look at console logging on desktop, which is where most developers are).

Yes, it makes sense to put all the logging in either one or the other. It's hard to imagine the current situation where `touchstart` and `touchmove` are in one, and `touchend` is in the other. It makes no sense. The trouble with `touchmove` is it fires so often that the log is swamped, but I don't suppose that really matters.

I think I'm generally pro in-page logging.

> There must be a way to make the log fixed length and scroll, so I'd do that.

You should be able to fix the log layout problem with some CSS like:

```css
#log {
  height: 100px;
  overflow: scroll;
}
```

> I am slightly tempted to remove the intialize button too. We added it to make simulation case easier - but you don't "need" it for the example, and it isn't hard to say "just reload the page".

Yes, I agree with that.

> P.S. I did update the logs to use template literals instead, so at least it isn't quite so old fashioned.

Also if we keep the in-page logging please remove the innerHTML:

```js
function log(msg) {
  const container = document.getElementById('log');
  container.textContent = `${ msg } \n${ container.textContent }`;
}
```

...works just as well, and we do have an open issue about innerHTML: https://github.com/mdn/content/issues/13496.

hamishwillee commented:
@wbamberg Re https://github.com/mdn/content/pull/13993#issuecomment-1074690620 - sure! I'm back on Friday and will look at this then. Much appreciated.

hamishwillee committed:
Remove the intialize button

hamishwillee committed:
Move all logging into the log windows

hamishwillee committed:
Replace innerHTML with textContent

hamishwillee commented:
@wbamberg All done, as discussed 
- logs kept but moved into the log container (some minor changes made to where logging occurs).
- write to log using textContent rather than innerHTML.
-  Remove initialize button

wbamberg reviewed (approved):
👍 thank you @hamishwillee ! This is so much better in so many ways.
